{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "816e157f",
   "metadata": {},
   "source": [
    "# FINDING SHORT TANDEM REPEATS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "18f62a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fasta file was read with UGENE\n",
    "#content from fasta was cleaned using sequence_cleaner (http://www.cellbiol.com/scripts/cleaner/dna_protein_sequence_cleaner.php)\n",
    "#read the text file containing the sequence and print it as a string\n",
    "with open('mtDapma3clones_MH683636_1.txt', 'r') as sequence:\n",
    "    seq = sequence.readlines()\n",
    "    seq = ''.join(seq)\n",
    "    #print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df66701",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample for debugging\n",
    "#seq = 'ATGATGATGATGATATAT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2fd8248",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28db0635",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a dictionary for combinations of nucleotides that could appear as a repeat; di and tri\n",
    "repeatsDict = {\n",
    "    \"dinucleotides\": [\"AT\",\"TA\", \"CG\", \"GC\", \"AA\", \"TT\",\n",
    "    \"CC\", \"GG\", \"AC\", \"AG\", \"TC\", \"TG\", \"CA\", \"CT\", \"GA\", \"GT\"],\n",
    "    \n",
    "    \"trinucleotides\": [\"AAA\",\"AAT\",\"AAC\",\"AAG\",\"ATG\",\"ATC\",\"ATA\",\n",
    "    \"ATT\",\"ACA\",\"ACT\",\"ACC\",\"ACG\",\"AGA\",\"AGT\",\"AGC\",\"AGG\",\"TTT\",\n",
    "    \"TTC\",\"TTA\",\"TTG\",\"TCT\",\"TCC\", \"TCA\",\"TCG\",\"TAT\",\"TAC\",\"TGT\", \n",
    "    \"TGC\",\"TGG\",\"CTT\",\"CTC\",\"CTA\",\"CTG\",\"CCT\", \"CCC\", \"CCA\",\"CCG\",\n",
    "    \"CAT\",\"CAC\",\"CAA\", \"CAG\", \"CGT\",\"CGC\",\"CGA\",\"CGG\",\"GTT\",\"GTC\",\n",
    "    \"GTA\",\"GTG\", \"GCT\",\"GCC\",\"GCA\",\"GCG\",\"GAT\",\"GAC\",\"GAA\",\"GAG\",\n",
    "    \"GGT\",\"GGC\",\"GGA\",\"GGG\"],\n",
    "    \n",
    "    \"stop\": [\"TAG\",\"TAA\",\"TGA\"]\n",
    "    }\n",
    "\n",
    "#print(repeatsDict[\"stop\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f379042d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c3768fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import library with built-in function of finding mutliple substrings in a string\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c74d48c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import library with built-in function of calculating binomial distribution\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ee2801",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2316f7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#BINOMIAL DISTRIBUTION\n",
    "# binom(k, n, p)\n",
    "#Define the number of successes (k), number of trials (n), and expected probability success (p).\n",
    "# k = total STRs; total\n",
    "# n = len of seq / 2 or 3; len(seq)/len(diN/tri/stC)... enter the diN/triN/stC as n\n",
    "# p = freqX of each possible STR; freqX(diN/triN/stC)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c4a1648f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#frequencey of nucleotides (Pb)\n",
    "def Pb(nucl):\n",
    "    return seq.count(nucl)/len(seq)\n",
    "\n",
    "fA = Pb('A')\n",
    "fT = Pb('T')\n",
    "fG = Pb('G')\n",
    "fC = Pb('C')\n",
    "#print(fA, fT, fG, fC)\n",
    "#print(fA+fT+fG+fC)\n",
    "\n",
    "def freqX(STRs): \n",
    "    x = 1\n",
    "    for i in STRs:\n",
    "        #print(i)\n",
    "        if i == 'A':\n",
    "            x *= fA\n",
    "        elif i == 'T':\n",
    "            x *= fT\n",
    "        elif i == 'G':\n",
    "            x *= fG\n",
    "        elif i == 'C':\n",
    "            x *= fC\n",
    "    #print(STRs, \"expected freq:\", x)\n",
    "    return x\n",
    "\n",
    "#freqX('TT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4edd969b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nTrials(STRs):\n",
    "    n = len(seq)/len(STRs)\n",
    "    return int(n)\n",
    "\n",
    "#nTrials('TGA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9790d4e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def binom(STRs, k, n, p):\n",
    "    statBD = stats.binom.pmf(k, n, p)\n",
    "    print( STRs, \"binomial distribution, Pb:\", statBD)\n",
    "    \n",
    "#binom('TGA', total, nTrials('TGA'), freqX('TGA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb99dd8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "60215ccc",
   "metadata": {},
   "source": [
    "# RESULTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "649d0a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a dictionary for all matching dinucleotides found and print position of all occurences of each dinucleotides\n",
    "diNucl = {}\n",
    "for i in repeatsDict[\"dinucleotides\"]:\n",
    "    match = re.finditer(pattern=i, string=seq)\n",
    "    diNucl[i] = [index.start() for index in match]\n",
    "#print(diNucl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7896e538",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 STRs ending at pos. 2361\n",
      "3 STRs ending at pos. 4132\n",
      "3 STRs ending at pos. 4585\n",
      "3 STRs ending at pos. 7029\n",
      "3 STRs ending at pos. 11803\n",
      "3 STRs ending at pos. 14809\n",
      "AT total STRs in the sequence : 18\n",
      "AT binomial distribution, Pb: 0.0\n",
      "\n",
      "3 STRs ending at pos. 664\n",
      "3 STRs ending at pos. 2360\n",
      "3 STRs ending at pos. 7328\n",
      "3 STRs ending at pos. 7962\n",
      "3 STRs ending at pos. 10978\n",
      "3 STRs ending at pos. 12360\n",
      "3 STRs ending at pos. 13413\n",
      "3 STRs ending at pos. 13673\n",
      "3 STRs ending at pos. 14603\n",
      "3 STRs ending at pos. 14810\n",
      "TA total STRs in the sequence : 30\n",
      "TA binomial distribution, Pb: 0.0\n",
      "\n",
      "3 STRs ending at pos. 417\n",
      "3 STRs ending at pos. 460\n",
      "3 STRs ending at pos. 673\n",
      "3 STRs ending at pos. 5820\n",
      "3 STRs ending at pos. 6491\n",
      "3 STRs ending at pos. 7009\n",
      "3 STRs ending at pos. 7315\n",
      "3 STRs ending at pos. 8019\n",
      "3 STRs ending at pos. 8425\n",
      "3 STRs ending at pos. 9020\n",
      "3 STRs ending at pos. 9218\n",
      "4 STRs ending at pos. 9295\n",
      "3 STRs ending at pos. 9422\n",
      "4 STRs ending at pos. 9485\n",
      "3 STRs ending at pos. 9702\n",
      "3 STRs ending at pos. 9844\n",
      "4 STRs ending at pos. 10254\n",
      "3 STRs ending at pos. 12425\n",
      "3 STRs ending at pos. 13039\n",
      "3 STRs ending at pos. 14282\n",
      "3 STRs ending at pos. 14619\n",
      "3 STRs ending at pos. 14997\n",
      "3 STRs ending at pos. 15136\n",
      "AA total STRs in the sequence : 72\n",
      "AA binomial distribution, Pb: 6.259253754370236e-287\n",
      "\n",
      "3 STRs ending at pos. 1116\n",
      "4 STRs ending at pos. 1831\n",
      "3 STRs ending at pos. 2072\n",
      "3 STRs ending at pos. 2341\n",
      "3 STRs ending at pos. 2630\n",
      "3 STRs ending at pos. 2871\n",
      "3 STRs ending at pos. 2919\n",
      "4 STRs ending at pos. 3305\n",
      "3 STRs ending at pos. 3794\n",
      "3 STRs ending at pos. 3895\n",
      "3 STRs ending at pos. 4117\n",
      "3 STRs ending at pos. 5168\n",
      "3 STRs ending at pos. 5911\n",
      "3 STRs ending at pos. 5930\n",
      "3 STRs ending at pos. 6017\n",
      "3 STRs ending at pos. 6217\n",
      "3 STRs ending at pos. 6988\n",
      "3 STRs ending at pos. 7396\n",
      "3 STRs ending at pos. 8174\n",
      "4 STRs ending at pos. 8243\n",
      "4 STRs ending at pos. 9992\n",
      "3 STRs ending at pos. 10245\n",
      "3 STRs ending at pos. 10322\n",
      "3 STRs ending at pos. 10440\n",
      "3 STRs ending at pos. 10783\n",
      "3 STRs ending at pos. 11047\n",
      "3 STRs ending at pos. 11108\n",
      "3 STRs ending at pos. 11334\n",
      "3 STRs ending at pos. 12732\n",
      "4 STRs ending at pos. 13901\n",
      "3 STRs ending at pos. 13909\n",
      "3 STRs ending at pos. 14275\n",
      "3 STRs ending at pos. 14561\n",
      "3 STRs ending at pos. 15100\n",
      "TT total STRs in the sequence : 107\n",
      "TT binomial distribution, Pb: 1.6323159085159326e-246\n",
      "\n",
      "3 STRs ending at pos. 380\n",
      "3 STRs ending at pos. 1948\n",
      "3 STRs ending at pos. 2019\n",
      "3 STRs ending at pos. 8823\n",
      "3 STRs ending at pos. 8911\n",
      "3 STRs ending at pos. 11905\n",
      "3 STRs ending at pos. 11983\n",
      "3 STRs ending at pos. 12234\n",
      "CC total STRs in the sequence : 24\n",
      "CC binomial distribution, Pb: 6.369412043764192e-69\n",
      "\n",
      "3 STRs ending at pos. 2301\n",
      "3 STRs ending at pos. 2683\n",
      "3 STRs ending at pos. 11380\n",
      "GG total STRs in the sequence : 9\n",
      "GG binomial distribution, Pb: 1.4577586981191431e-64\n",
      "\n",
      "3 STRs ending at pos. 708\n",
      "3 STRs ending at pos. 7746\n",
      "3 STRs ending at pos. 8123\n",
      "3 STRs ending at pos. 14116\n",
      "3 STRs ending at pos. 15033\n",
      "AC total STRs in the sequence : 15\n",
      "AC binomial distribution, Pb: 5.803125182628009e-172\n",
      "\n",
      "3 STRs ending at pos. 344\n",
      "3 STRs ending at pos. 6908\n",
      "3 STRs ending at pos. 9586\n",
      "3 STRs ending at pos. 10373\n",
      "AG total STRs in the sequence : 12\n",
      "AG binomial distribution, Pb: 3.8375757719166864e-153\n",
      "\n",
      "3 STRs ending at pos. 1588\n",
      "3 STRs ending at pos. 2025\n",
      "3 STRs ending at pos. 2393\n",
      "3 STRs ending at pos. 9238\n",
      "TC total STRs in the sequence : 12\n",
      "TC binomial distribution, Pb: 2.2374729512669924e-175\n",
      "\n",
      "3 STRs ending at pos. 7747\n",
      "CA total STRs in the sequence : 3\n",
      "CA binomial distribution, Pb: 1.024345275116083e-192\n",
      "\n",
      "3 STRs ending at pos. 1099\n",
      "4 STRs ending at pos. 2026\n",
      "CT total STRs in the sequence : 7\n",
      "CT binomial distribution, Pb: 9.330232850784228e-184\n",
      "\n",
      "3 STRs ending at pos. 5840\n",
      "3 STRs ending at pos. 6909\n",
      "3 STRs ending at pos. 8432\n",
      "3 STRs ending at pos. 9585\n",
      "3 STRs ending at pos. 13990\n",
      "GA total STRs in the sequence : 15\n",
      "GA binomial distribution, Pb: 9.936686932983595e-149\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#for each item in the dinucleotides, find the positions showing at least 3 repeats \n",
    "for dn, diN in zip(diNucl.values(), diNucl.keys()):\n",
    "    #print(diN)  \n",
    "    \n",
    "    total = 0\n",
    "    count = 0\n",
    "    x = 0\n",
    "    # extend length by 2 to allow entire element to run, this is just a random number\n",
    "    dn.extend((-7, -7))\n",
    "    # iterate through the length of the selected dinucleotide\n",
    "    while x < len(dn)-2:\n",
    "        x += 1\n",
    "        \n",
    "        #for each item in the list of the selected dinucleotide\n",
    "        for i in dn:\n",
    "            # select for next 2, so that it can be identified as at least 3 repeats\n",
    "            if i == dn[x]-2 and i == dn[x+1]-4:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "\n",
    "            # else check next 1 and preceeding 1   \n",
    "            elif i == dn[x]-2 and i == dn[x-2]+2:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "\n",
    "            # else check preceeding 2   \n",
    "            elif i == dn[x-2]+2 and i == dn[x-3]+4:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "                print(count, 'STRs ending at pos.', i+2)\n",
    "                \n",
    "                # add count to total, then reset counter \n",
    "                total += count\n",
    "                count = 0\n",
    "     \n",
    "    # print only nucleotides with repeats\n",
    "    if total > 0:\n",
    "        print(diN, 'total STRs in the sequence :', total)\n",
    "        \n",
    "        # print binomial probability\n",
    "        binom(diN, total, nTrials(diN), freqX(diN))\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9453b40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41427d81",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a dictionary for all matching trinucleotides found and print position of all occurences of each trinucleotides\n",
    "triNucl = {}\n",
    "for i in repeatsDict[\"trinucleotides\"]:\n",
    "    match = re.finditer(pattern=i, string=seq)\n",
    "    triNucl[i] = [index.start() for index in match]\n",
    "#print(triNucl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71792f1c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 STRs ending at pos. 10255\n",
      "AAA total STRs in the sequence : 3\n",
      "AAA binomial distribution, Pb: 2.1859668044193605e-80\n",
      "\n",
      "3 STRs ending at pos. 817\n",
      "3 STRs ending at pos. 1048\n",
      "AAT total STRs in the sequence : 6\n",
      "AAT binomial distribution, Pb: 4.669242549912849e-75\n",
      "\n",
      "4 STRs ending at pos. 818\n",
      "3 STRs ending at pos. 8687\n",
      "ATA total STRs in the sequence : 7\n",
      "ATA binomial distribution, Pb: 1.3340228135702085e-73\n",
      "\n",
      "3 STRs ending at pos. 4724\n",
      "TCT total STRs in the sequence : 3\n",
      "TCT binomial distribution, Pb: 7.30178260091153e-39\n",
      "\n",
      "3 STRs ending at pos. 7695\n",
      "TAT total STRs in the sequence : 3\n",
      "TAT binomial distribution, Pb: 2.2231222767916222e-79\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#for each item in the trinucleotides,  find the positions showing at least 3 repeats\n",
    "for tn, triN in zip(triNucl.values(), triNucl.keys()):\n",
    "    #print(triN)  \n",
    "    \n",
    "    total = 0\n",
    "    count = 0\n",
    "    x = 0\n",
    "    # extend length by 2 values to allow entire element to run, this is just a random number\n",
    "    tn.extend((-7, -7))\n",
    "    # iterate through the length of the selected trinucleotides\n",
    "    while x < len(tn)-2:\n",
    "        x += 1\n",
    "                \n",
    "        #for each item in the list of the selected trinucleotides\n",
    "        for i in tn:\n",
    "            # select for next 2, so that it can be identified as at least 3 repeats\n",
    "            if i == tn[x]-3 and i == tn[x+1]-6:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "            \n",
    "            # else check next 1, and preceeding 1   \n",
    "            elif i == tn[x]-3 and i == tn[x-2]+3:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "                \n",
    "            # else check preceeding 2\n",
    "            elif i == tn[x-2]+3 and i == tn[x-3]+6:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "                print(count, 'STRs ending at pos.', i+3)\n",
    "                \n",
    "                # add count to total, then reset counter \n",
    "                total += count\n",
    "                count = 0\n",
    "    \n",
    "    # print only nucleotides with repeats\n",
    "    if total > 0:\n",
    "        print(triN, 'total STRs in the sequence :', total, )\n",
    "        \n",
    "        # print binomial probability\n",
    "        binom(triN, total, nTrials(triN), freqX(triN))\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaa8eeda",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "153913e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#create a dictionary for all matching stop codons found and print position of all occurences of each stop codons\n",
    "stopCodons = {}\n",
    "for i in repeatsDict[\"stop\"]:\n",
    "    match = re.finditer(pattern=i, string=seq)\n",
    "    stopCodons[i] = [index.start() for index in match]\n",
    "#print(stopCodons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "de13d5f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 STRs ending at pos. 819\n",
      "3 STRs ending at pos. 1047\n",
      "3 STRs ending at pos. 8688\n",
      "TAA total STRs in the sequence : 10\n",
      "TAA binomial distribution, Pb: 1.4803323638208016e-69\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#for each item in the stop codons,  find the positions showing at least 3 repeats\n",
    "for sc, stC in zip(stopCodons.values(), stopCodons.keys()):\n",
    "    #print(stC)  \n",
    "    \n",
    "    total = 0\n",
    "    count = 0\n",
    "    x = 0\n",
    "    \n",
    "    # extend length by 2 values to allow entire element to run, this is just a random number\n",
    "    sc.extend((-7,-7))\n",
    "    # iterate through the length of the selected stop codon\n",
    "    while x < len(sc)-2:\n",
    "        x += 1\n",
    "                \n",
    "        #for each item in the list of the selected stop codon\n",
    "        for i in sc:\n",
    "            # select for next 2, so that it can be identified as at least 3 repeats\n",
    "            if i == sc[x]-3 and i == sc[x+1]-6:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "            \n",
    "            # else check next 1, and preceeding 1   \n",
    "            elif i == sc[x]-3 and i == sc[x-2]+3:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "                \n",
    "            # else check preceeding 2\n",
    "            elif i == sc[x-2]+3 and i == sc[x-3]+6:\n",
    "                count += 1\n",
    "                #print(i, count)\n",
    "                print(count, 'STRs ending at pos.', i+3)\n",
    "                \n",
    "                # add count to total, then reset counter \n",
    "                total += count\n",
    "                count = 0\n",
    "                \n",
    "    # print only nucleotides with repeats\n",
    "    if total > 0:\n",
    "        print(stC, 'total STRs in the sequence :', total)\n",
    "        \n",
    "        # print binomial probability\n",
    "        binom(stC, total, nTrials(stC), freqX(stC))\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e5a1b99",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
